base/iterators.jl

```diff
@@ -1128,6 +1128,7 @@ end
 reverse(p::ProductIterator) = ProductIterator(Base.map(reverse, p.iterators))
 last(p::ProductIterator) = Base.map(last, p.iterators)
 intersect(a::ProductIterator, b::ProductIterator) = ProductIterator(intersect.(a.iterators, b.iterators))
+getindex(p::ProductIterator, inds...) = map(getindex, p.iterators, inds)
```

Comment:
So here is an actual problem with this implementation:

```julia
julia> a = Iterators.product([1 2; 3 4], 1:2);

julia> b = collect(a)
2×2×2 Array{Tuple{Int64, Int64}, 3}:
[:, :, 1] =
 (1, 1)  (2, 1)
 (3, 1)  (4, 1)

[:, :, 2] =
 (1, 2)  (2, 2)
 (3, 2)  (4, 2)

julia> b[1,1,1]
(1, 1)

julia> map(getindex, a.iterators, (1,1,1))
ERROR: BoundsError: attempt to access Tuple{} at index [1]
Stacktrace:
 [1] getindex(t::Tuple, i::Int64)
   @ Base ./tuple.jl:29
 [2] map (repeats 3 times)
   @ ./tuple.jl:250 [inlined]
 [3] top-level scope
   @ REPL[24]:1
```

i.e. map assumes every iterator is one dimensional. The following should be a fix I think

```julia
function getindex(prod::ProductIterator, indices...)
    return _prod_getindex(prod.iterators, indices...)
end
_prod_getindex(::Tuple{}) = ()
function _prod_getindex(p_vecs::Tuple, indices...)
    v = first(p_vecs)
    n = ndim(v)
    return (
        v[indices[1:n]...],
        _prod_getindex(Base.tail(p_vecs), indices[n+1:end]...)...
    )
end
```

Reply:
you also need to manage all types of access variants (colons, ranges, ...) if you don't limit indices to `Vararg{Int,N}`. But for this you need to determine `N` (i.e. `ndims`) statically.